### PR TITLE
Peeling union find with boundaries

### PR DESCRIPTION
**Pull request by quantumgizmos**
Enables decoding of surface codes with the peeling version of union find.

### CODE REVIEW

File: src_cpp/union_find.hpp [hunk 1]
```diff
@@ -523,15 +596,38 @@ class UfDecoder{
...
+                }
+            }
+
+
         }
 
         std::vector<uint8_t>& peel_decode(const std::vector<uint8_t>& syndrome, const std::vector<double>& bit_weights = NULL_DOUBLE_VECTOR, int bits_per_step = 1){
 
+            if(!this->is_planar_code){
+                throw(std::runtime_error("Peel decoder only works for planar codes. Use the matrix_decode method for more general codes."));
```
Comment (timohillmann) [inline src_cpp/union_find.hpp:627]: Why does it not work for non-planar codes? It should work for the Toric code.

File: src_cpp/union_find.hpp [hunk 2]
```diff
@@ -557,7 +653,7 @@ class UfDecoder{
 
                 invalid_clusters.clear();
                 for(auto cl: clusters){
-                    if(cl->active == true && cl->parity() == 1){
+                    if(cl->active == true && cl->parity() == 1 && cl->contains_boundary_bits == false){
```
Comment (timohillmann) [inline src_cpp/union_find.hpp:656]: So you turn off a cluster whenever it grows into a boundary. Are there cases and UF grow orders for which this is suboptimal (or better) than growing until it becomes valid, i.e., merging with another invalid cluster? I am almost certain there are multiple cases where the resulting clusters are not equivalent.

File: src_cpp/union_find.hpp [hunk 3]
```diff
@@ -241,14 +255,28 @@ struct Cluster{
...
         }
 
+
+
         int check_neighbours[2];
         for(int bit_index: this->bit_nodes){
             check_neighbours[0] = this->pcm.column_heads[bit_index]->up->row_index;
             check_neighbours[1] = this->pcm.column_heads[bit_index]->down->row_index;
+
+            if(check_neighbours[0] == check_neighbours[1]){
```
Comment (timohillmann) [inline src_cpp/union_find.hpp:276]: This handles the case of column weight 1 bits, right?

File: src_cpp/union_find.hpp (hunk 1 above)
Comment (quantumgizmos) [reply src_cpp/union_find.hpp:627]: What's the general name for a code that produces pair-like or pair to boundary excitations.

File: src_cpp/union_find.hpp (hunk 2 above)
Comment (quantumgizmos) [reply src_cpp/union_find.hpp:656]: As soon as a cluster intersects a boundary, I mark it as "valid". I.e. it no longer grows. However, other clusters can grow into it and merge.

File: src_cpp/union_find.hpp (hunk 3 above)
Comment (quantumgizmos) [reply src_cpp/union_find.hpp:276]: Yes precisely. And adds a virtual boundary node.

File: src_cpp/union_find.hpp (hunk 2 above)
Comment (quantumgizmos) [reply src_cpp/union_find.hpp:656]: As for whether this is optimal ... probably not. I've simply decided to treat boundaries as equivalent to defects in the lattice. Hence the decision to freeze growth once the boundary is hit.

Comment (quantumgizmos) [reply src_cpp/union_find.hpp:656]: Thinking about this some more, shouldn't it generally be the case that we want to keep our clusters as small as possible.?

File: src_cpp/union_find.hpp (hunk 1 above)
Comment (timohillmann) [reply src_cpp/union_find.hpp:627]: any code with max bit_degree 2 should do that. Planar to me means non-periodic.

File: src_cpp/union_find.hpp (hunk 2 above)
Comment (timohillmann) [reply src_cpp/union_find.hpp:656]: Thinking about it more, I am not sure it makes a difference, yes.

File: src_cpp/union_find.hpp [hunk 4]
```diff
@@ -21,6 +21,7 @@
 namespace ldpc::uf{
 
 const std::vector<double> NULL_DOUBLE_VECTOR = {};
+tsl::robin_set<int> NULL_INT_ROBIN_SET = {};
```
[the remark this answers is not in the document]
Comment (quantumgizmos) [reply src_cpp/union_find.hpp:24]: Yes I agree. That makes more sense

File: cpp_test/TestUnionFind.cpp [hunk 5]
```diff
@@ -113,13 +114,100 @@ TEST(UfDecoder, ring_code3){
...
...nsert_entry(6,6);pcm.insert_entry(6,11);pcm.insert_entry(6,29);pcm.insert_entry(6,30);pcm.insert_entry(7,7);pcm.insert_entry(7,12);pcm.insert_entry(7,30);pcm.insert_entry(7,31);pcm.insert_entry(8,8);pcm.insert_entry(8,13);pcm.insert_entry(8,31);pcm.insert_entry(8,32);pcm.insert_entry(9,9);pcm.insert_entry(9,14);pcm.insert_entry(9,32);pcm.insert_entry(10,10);pcm.insert_entry(10,15);pcm.insert_entry(10,33);pcm.insert_entry(11,11);pcm.insert_entry(11,16);pcm.insert_entry(11,33);pcm.insert_entry(11,34);pcm.insert_entry(12,12);pcm.insert_entry(12,17);pcm.insert_entry(12,34);pcm.insert_entry(12,35);pcm.insert_entry(13,13);pcm.insert_entry(13,18);pcm.insert_entry(13,35);pcm.insert_entry(13,36);pcm.insert_entry(14,14);pcm.insert_entry(14,19);pcm.insert_entry(14,36);pcm.insert_entry(15,15);pcm.insert_entry(15,20);pcm.insert_entry(15,37);pcm.insert_entry(16,16);pcm.insert_entry(16,21);pcm.insert_entry(16,37);pcm.insert_entry(16,38);pcm.insert_entry(17,17);pcm.insert_entry(17,22);pcm.insert_entry(17,38);pcm.insert_entry(17,39);pcm.insert_entry(18,18);pcm.insert_entry(18,23);pcm.insert_entry(18,39);pcm.insert_entry(18,40);pcm.insert_entry(19,19);pcm.insert_entry(19,24);pcm.insert_entry(19,40);
```
[the remark this answers is not in the document]
Comment (quantumgizmos) [reply cpp_test/TestUnionFind.cpp:182]: Hi. This was just a specific test from the Python Monte Carlo tests that was failing. I ported the pcm and the syndrome over to CPP for debugging. It seems to be resolved now so this test can be deleted.

**Pull request merged**